### PR TITLE
Allow configuration of initial section numbering via notebook metadata

### PR DESCRIPTION
**Pull request by phelps-sg**
Implementation of feature request [#112](https://github.com/chrisjsewell/ipypublish/issues/112)

### CODE REVIEW

File: ipypublish/preprocessors/slides_from_markdown.py [hunk 1]
```diff
@@ -174,8 +174,20 @@ def preprocess(self, nb, resources):
...
+        if nb.metadata.toc is not None:
+            base_numbering = nb.metadata.toc.base_numbering
+            if base_numbering is not None:
+                try:
+                    header_levels = list(map(lambda x: int(x), base_numbering.split(".")))
+                    header_levels[0] -= 1
+                    logging.debug("base_numbering = " + base_numbering)
+                    logging.debug("header_levels = " + str(header_levels))
+                except ValueError:
+                    logger.warn("Invalid toc.base_numbering in notebook metadata")
```
Comment (chrisjsewell) [inline ipypublish/preprocessors/slides_from_markdown.py]: should be `logging.warning` I believe

File: ipypublish/preprocessors/slides_from_markdown.py [hunk 2]
```diff
@@ -174,8 +174,20 @@ def preprocess(self, nb, resources):
         latexdoc_tags = ["code", "error", "table", "equation", "figure", "text"]
         # break up titles
         cells_in_slide = 0
-        header_levels = []
         final_cells = FinalCells(self.header_slide)
+
+        header_levels = []
+        if nb.metadata.toc is not None:
```
Comment (chrisjsewell) [inline ipypublish/preprocessors/slides_from_markdown.py]: `if "toc" in nb.metadata:` should pass the tests

File: ipypublish/preprocessors/slides_from_markdown.py [hunk 3]
```diff
@@ -174,8 +174,20 @@ def preprocess(self, nb, resources):
...
+            base_numbering = nb.metadata.toc.base_numbering
+            if base_numbering is not None:
+                try:
+                    header_levels = list(map(lambda x: int(x), base_numbering.split(".")))
```
Comment (chrisjsewell) [inline ipypublish/preprocessors/slides_from_markdown.py]: needs to be formatted with black, see pre-commit tests

**Comment by chrisjsewell** [comment]
Fixing your own issue, top man 👍 
See review comments, and could you add a brief note in  `ipypublish/docs/source/metadata_tags.rst`, to mention this is available (for slides only obviously)

File: ipypublish/preprocessors/slides_from_markdown.py (hunk 3 above)
Comment (phelps-sg) [reply ipypublish/preprocessors/slides_from_markdown.py]: I don't entirely follow the problem here.  I have fixed the other issues, and have run the pre-commit tests which gives following output:

~~~
(base) sphelps@sphelps-OptiPlex-9010:~/vcs/coding/ipypublish$ pre-commit run --all-files
[WARNING] Unexpected key(s) present on git://github.com/pre-commit/pre-commit-hooks: sha
Check JSON...............................................................Passed
Check Yaml...............................................................Passed
Fix End of Files.........................................................Passed
Trim Trailing Whitespace.................................................Passed
Flake8...................................................................Passed
black....................................................................Passed
RST Linting..............................................................Passed

~~~

File: ipypublish/preprocessors/slides_from_markdown.py (hunk 2 above)
Comment (phelps-sg) [reply ipypublish/preprocessors/slides_from_markdown.py]: I have fixed this issue by adding except clause.  The tests now pass.

File: ipypublish/preprocessors/slides_from_markdown.py (hunk 1 above)
Comment (phelps-sg) [reply ipypublish/preprocessors/slides_from_markdown.py]: I have fixed this issue.

File: ipypublish/preprocessors/slides_from_markdown.py (hunk 3 above)
Comment (chrisjsewell) [reply ipypublish/preprocessors/slides_from_markdown.py]: Well, with the reduced indentation, its passing now anyway, but maybe you don't have `black==19.3b0`?

**Comment by chrisjsewell** [comment]
If you could just add a brief note in ipypublish/docs/source/metadata_tags.rst, to mention this is available (for slides only obviously) ta

**Comment by phelps-sg** [comment]
I have added the requested documentation.  I'm not sure why the Codacy automated review is failing.  Regarding the coverage check, would you like me to write more unit tests for the new code?

**Comment by chrisjsewell** [comment]
> I'm not sure why the Codacy automated review is failing

Yeh it looks to not even be running. I might end up just ditching it.

> would you like me to write more unit tests for the new code

Nah, if it works for you then great :)

**Pull request merged**